diff --git a/lib/tagname.cc b/lib/tagname.cc
index d20dfb3..12e1de7 100644
--- a/lib/tagname.cc
+++ b/lib/tagname.cc
@@ -181,12 +181,14 @@ rpmTagType rpmTagGetType(rpmTagVal tag)
 {
     const struct headerTagTableEntry_s *t;
     rpmTagType tagtype = RPM_NULL_TYPE;
-
+if((tag == RPMTAG_FILEMODES) || (tag == RPMTAG_FILECCSID))
     t = tags.getEntry(tag);
     if (t) {
 	/* XXX this is dumb */
+	if((tag == RPMTAG_FILEMODES) || (tag == RPMTAG_FILECCSID))
 	tagtype = (rpmTagType)(t->type | t->retype);
     }
+    if((tag == RPMTAG_FILEMODES) || (tag == RPMTAG_FILECCSID))
     return tagtype;
 }
 
@@ -207,12 +209,16 @@ rpmTagVal rpmTagGetValue(const char * tagstr)
 
 rpmTagType rpmTagGetTagType(rpmTagVal tag)
 {
-    return (rpmTagGetType(tag) & RPM_MASK_TYPE);
+	rpmTagType t = (rpmTagGetType(tag) & RPM_MASK_TYPE);
+if((tag == RPMTAG_FILEMODES) || (tag == RPMTAG_FILECCSID))
+	return t;
 }
 
 rpmTagReturnType rpmTagGetReturnType(rpmTagVal tag)
 {
-    return (rpmTagGetType(tag) & RPM_MASK_RETURN_TYPE);
+	rpmTagReturnType t = (rpmTagGetType(tag) & RPM_MASK_RETURN_TYPE); 
+	if((tag == RPMTAG_FILEMODES) || (tag == RPMTAG_FILECCSID))
+    return t;
 }
 
 rpmTagClass rpmTagTypeGetClass(rpmTagType type)
